Avoid UB when indexing into size-0 tensors

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65878



If we attempt to compute an offset into an empty tensor we trigger UB, since
we'd be adding an offset to a nullptr, which is UB
(https://reviews.llvm.org/D67122) even if we never use the pointer.

Since indexing into an empty tensor yields an empty tensor anyways, let's just
return the underlying (null) data ptr in this case.

Differential Revision: [D31296914](https://our.internmc.facebook.com/intern/diff/D31296914)

bertmaher committed:
Avoid UB when indexing into size-0 tensors

When setting up a TensorIterator we cache the `data_ptr`s, but in the
case of a tensor that indexes into a size-0 tensor it's not valid to access
`data_ptr`, since we'd be adding an offset to a nullptr, which is UB
(https://reviews.llvm.org/D67122) even if we never use the pointer.

Let's only cache the data_ptr if the tensor actually has elements.

Differential Revision: [D31296914](https://our.internmc.facebook.com/intern/diff/D31296914/)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -1358,7 +1358,11 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
-    op.data = op.tensor->data_ptr();
+    // Don't access the data_ptr if the tensor is size-0; if there's an offset
+    // into storage that will trigger UB.
+    if (op.tensor->numel() != 0) {
```
doesn't look like the right place for the fix, the problem is in https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1183. I was hoping that tensors without storage don't have storage_offset_, but that's not true for slices.

bertmaher replied:
> We could check for is_empty() instead of storage_offset_ there

This seems reasonable to me.  It's in keeping with the semantics of the function  (an index into an empty tensor is still empty, after all), whereas quietly forcing storage_offset_ to 0 seems a bit more "magical".

ngimel replied:
This lgtm, @ezyang if you can take a quick look would be great

bertmaher committed:
Put the fix in TensorImpl instead of TensorIterator on "Avoid UB when indexing into size-0 tensors"




If we attempt to compute an offset into an empty tensor we trigger UB, since
we'd be adding an offset to a nullptr, which is UB
(https://reviews.llvm.org/D67122) even if we never use the pointer.

Since indexing into an empty tensor yields an empty tensor anyways, let's just
return the underlying (null) data ptr in this case.


Differential Revision: [D31296914](https://our.internmc.facebook.com/intern/diff/D31296914/)

Differential Revision: [D31296914](https://our.internmc.facebook.com/intern/diff/D31296914)

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -1180,13 +1180,15 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+    // tensor's storage will be nullptr, and adding a nonzero offset to nullptr
+    // is UB.  So we skip the offset computation in this case.
+    if (is_empty()) {
       return storage_.data();
```
Why not just return nullptr directly then

bertmaher replied:
Sure, done.

bertmaher committed:
Just return nullptr on "Avoid UB when indexing into size-0 tensors"




If we attempt to compute an offset into an empty tensor we trigger UB, since
we'd be adding an offset to a nullptr, which is UB
(https://reviews.llvm.org/D67122) even if we never use the pointer.

Since indexing into an empty tensor yields an empty tensor anyways, let's just
return the underlying (null) data ptr in this case.

Differential Revision: [D31296914](https://our.internmc.facebook.com/intern/diff/D31296914)

[ghstack-poisoned]
